src/category_theory/preadditive/hom_orthogonal.lean

```diff
@@ -0,0 +1,195 @@
+and matrix entries in `i`-th block living in the endomorphisms of `s i`. -/
+@[simps] noncomputable
+def matrix_decomposition
+  (o : hom_orthogonal s) {α β : Type*} [fintype α] [fintype β] {f : α → ι} {g : β → ι} :
+  (⨁ (λ a, s (f a)) ⟶ ⨁ (λ b, s (g b))) ≃
+    Π (i : ι), matrix (g ⁻¹' {i}) (f ⁻¹' {i}) (End (s i)) :=
+{ to_fun := λ z i j k,
+    eq_to_hom (by { rcases k with ⟨k, ⟨⟩⟩, simp, }) ≫
+      biproduct.components z k j ≫ eq_to_hom (by { rcases j with ⟨j, ⟨⟩⟩, simp, }),
+  inv_fun := λ z, biproduct.matrix (λ j k, if h : f j = g k then
+      z (f j) ⟨k, by simp [h]⟩ ⟨j, by simp⟩ ≫ eq_to_hom (by simp [h])
+    else
+      0),
+  left_inv := λ z, begin
+    ext j k,
+    simp only [category.assoc, biproduct.lift_π, biproduct.ι_matrix],
+    split_ifs,
+    { simp, refl, },
+    { symmetry, apply o.eq_zero h, },
+  end,
+  right_inv := λ z, begin
+    ext i ⟨j, w⟩ ⟨k, ⟨⟩⟩,
+    simp only [set.mem_preimage, set.mem_singleton_iff],
+    simp [w.symm], refl,
+  end, }
+
+end
+
+section
+variables [preadditive C] [has_finite_biproducts C]
+
+/-- `hom_orthogonal.matrix_decomposition` as an additive equivalence. -/
+@[simps] noncomputable
+def matrix_decomposition_add_equiv
+  (o : hom_orthogonal s) {α β : Type*} [fintype α] [fintype β] {f : α → ι} {g : β → ι} :
+  (⨁ (λ a, s (f a)) ⟶ ⨁ (λ b, s (g b))) ≃+
+    Π (i : ι), matrix (g ⁻¹' {i}) (f ⁻¹' {i}) (End (s i)) :=
+{ map_add' := λ w z, by { ext, dsimp [biproduct.components], simp, },
+  ..o.matrix_decomposition, }.
+
+@[simp]
+lemma matrix_decomposition_id
+  (o : hom_orthogonal s) {α : Type*} [fintype α] {f : α → ι} (i : ι) :
+  o.matrix_decomposition (𝟙 (⨁ (λ a, s (f a)))) i = 1 :=
+begin
+  ext ⟨b, ⟨⟩⟩ ⟨a⟩,
+  simp only [set.mem_preimage, set.mem_singleton_iff] at j_property,
+  simp only [category.comp_id, category.id_comp, category.assoc, End.one_def, eq_to_hom_refl,
+    matrix.one_apply, hom_orthogonal.matrix_decomposition_apply, biproduct.components],
+  split_ifs with h,
+  { cases h, simp, },
+  { convert comp_zero,
+    simpa using biproduct.ι_π_ne _ (ne.symm h), },
+end
+
+lemma matrix_decomposition_comp
+  (o : hom_orthogonal s)
+  {α β γ : Type*} [fintype α] [fintype β] [fintype γ] {f : α → ι} {g : β → ι} {h : γ → ι}
+  (z : (⨁ (λ a, s (f a)) ⟶ ⨁ (λ b, s (g b)))) (w : (⨁ (λ b, s (g b)) ⟶ ⨁ (λ c, s (h c))))
+  (i : ι) :
+  o.matrix_decomposition (z ≫ w) i = o.matrix_decomposition w i ⬝ o.matrix_decomposition z i :=
+begin
+  ext ⟨c, ⟨⟩⟩ ⟨a⟩,
+  simp only [set.mem_preimage, set.mem_singleton_iff] at j_property,
+  simp only [matrix.mul_apply, limits.biproduct.components,
+    hom_orthogonal.matrix_decomposition_apply,
+    category.comp_id, category.id_comp, category.assoc, End.mul_def,
+    eq_to_hom_refl, eq_to_hom_trans_assoc, finset.sum_congr],
+  conv_lhs { rw ←category.id_comp w, rw ←biproduct.total, },
+  simp only [preadditive.sum_comp, preadditive.comp_sum],
+  apply finset.sum_congr_set,
+  { intros, simp, refl, },
+  { intros b nm, simp at nm,
```

Comment:
Noticed a non-terminal `simp` here.

(And, if you want, a few lines up there's a double `rw` you might be able to coalesce.)

Reply:
Done!